lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,300 @@
+            dispose();
+            reject(e);
+          }
+        }
+      }
+      if (res.statusCode > 303 || res.statusCode < 200) {
+        dispose();
+        reject(
+          new ERR_UNSUPPORTED_ESM_NETWORK_BAD_RESPONSE(
+            parsed.href,
+            'HTTP response returned status code of ' + res.statusCode));
+        return;
+      }
+      const { headers } = res;
+      const contentType = headers['content-type'];
+      if (!contentType) {
+        dispose();
+        reject(new ERR_UNSUPPORTED_ESM_NETWORK_BAD_RESPONSE(
+          parsed.href,
+          'content-type is required'));
+        return;
+      }
+      /**
+       * @type {CacheEntry}
+       */
+      const entry = {
+        resolvedHREF: parsed.href,
+        headers: {
+          'content-type': res.headers['content-type']
+        },
+        body: new Promise((f, r) => {
+          const buffers = [];
+          let size = 0;
+          let bodyStream = res;
+          let onError;
+          if (res.headers['content-encoding'] === 'br') {
+            bodyStream = createBrotliDecompress();
+            onError = function onError(e) {
+              bodyStream.close();
+              dispose();
+              reject(e);
+              r(e);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else if (res.headers['content-encoding'] === 'gzip' ||
+            res.headers['content-encoding'] === 'deflate') {
+            bodyStream = createUnzip();
+            onError = function onError(e) {
+              bodyStream.close();
+              dispose();
+              reject(e);
+              r(e);
+            };
+            res.on('error', onError);
+            res.pipe(bodyStream);
+          } else {
+            onError = function onError(e) {
+              dispose();
+              reject(e);
+              r(e);
+            };
+          }
+          bodyStream.on('error', onError);
+          bodyStream.on('data', (d) => {
+            ArrayPrototypePush(buffers, d);
+            size += d.length;
+          });
+          bodyStream.on('end', () => {
+            const body = entry.body = /** @type {Buffer} */(
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      getCache.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  getCache.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[') &&
+        StringPrototypeEndsWith(hostname, ']')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
```

Comment:
Should we skip the DNS lookup and only allow the bare IP form to avoid the DNS rebinding attack?

Reply:
in general only allowing IPs is unusable by most developers. DNS rebinding could be mitigated by a cache or external firewall but I do believe some form of DNS resolution is needed. I'd be open to any mitigation desired but think external firewall is reasonable personally.

Reply:
to clarify, various workflows use local aliases like `dev.mycompany.com` that are DNS entries that re-route to localhost in test environments.